Build and check CBMC proofs concurrently with litani

markrtuttle opened the pull request:
This pull request modifies the method of building and checking CBMC proofs to use litani.  

Litani is a build system based on ninja that allows us to build and check CBMC proofs concurrently, and to produce a summary that gives the current status of proof progress, and ends with an overall summary of the proof results.  Moving to a litani build also makes it possible to move to a new implementation of continuous integration.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Commit:
Add litani submodule to build CBMC proofs concurrently.

This commit adds Litani, hosted at https://github.com/awslabs/aws-build-accumulator
Litani is a build tool for running proofs and generating reports. This
commit is in preparation for converting the CBMC proofs to run using
Litani, which will help with proof and CI maintenance.

Commit:
Make CBMC patch.py look for touch file before patching.

Commit:
Fix CBMC MakefileLinux.json prerequisite symbol.

This commit fixes the symbol used to represent the prerequisites of a
Makefile target from $< to $^. The latter symbol expands to all of a
target's prerequisites, not just the first one. This is needed to ensure
that all object files get linked together.

Commit:
Make CBMC Makefile.common check for external sat solver.

Commit:
Make CBMC proofs build concurrently with litani.

karkhaz commented on tools/cbmc/proofs/Makefile.template:
```diff
@@ -73,14 +63,27 @@ unpatch:
-$(ENTRY)1.goto: $(OBJS)
-	$(GOTO_CC) @COMPILE_LINK@ @RULE_OUTPUT@ --function harness $(OBJS)
+queue_datastructure.h: gotos/$(FREERTOS)/freertos_kernel/queue.goto
+	python3 @TYPE_HEADER_SCRIPT@ --binary $(CWD)/gotos$(FREERTOS)/freertos_kernel/queue.goto --c-file $(FREERTOS)/freertos_kernel/queue.c
```
* There should be a `/` in `$(CWD)/gotos$(FREERTOS)` between `gotos` and `$(FREERTOS)`
* Currently, the prerequisite is `gotos/$(FREERTOS)/...` but the argument is `$(CWD)/gotos/...`. I would prefer if the argument was just `$@` so that we don't write the same thing twice and have it go out-of-date. I believe there's no need for `$(CWD)` because this command will be run in the proof directory, but IMO you should still have it there just so that an absolute path shows up in the log.

markrtuttle replied:
I have addressed all of these issues with `--binary @RULE_INPUT@`.  This required modifying the type header script to expand this relative path to an absolute patch before running the goto-instrument command that uses this path.  The script changes directory to a temporary directory before running goto-instrument, so an absolute path is required, probably explaining  the reason for the $(CWD) prefix used in the prior version of this Makefile.template.

karkhaz commented on tools/cbmc/proofs/Makefile.template:
```diff
@@ -110,11 +113,13 @@ $(ENTRY).goto: $(ENTRY)5.goto
 # Rules for running CBMC
 
 goto:
```
Do you not want to do the `_goto`, `_report`, etc trick that we do for other repositories, so that people can just run `make report` in the proof directory and have it 'just work' with litani?

markrtuttle replied:
No, we are not changing the targets themselves.  We are using litani to run `make goto`, we are not using `make goto` to run litani as we do in the starter kit.

karkhaz commented on tools/cbmc/proofs/run-cbmc-proofs.py:
```diff
@@ -0,0 +1,313 @@
+
+    if not args.no_standalone:
+        run_cmd(
+            ["./prepare.py"], check=True, cwd=str(proof_root))
```
Is this idempotent now?

markrtuttle replied:
Yes, the patch module that this script calls now checks for the patching touch file before patching.

Commit:
Make CBMC type header script use absolute paths.

The type header script runs goto-instrument in a temporary directory,
so the script must construct absolute paths to the inputs before
changing directory to run the command.
